### PR TITLE
[JENKINS-66771] Add a "conflictsWith" concept to agent RetentionStrategy…

### PR DESCRIPTION
**Pull request by jimklimov**
…(goes with inDemandDelay handling)

See [JENKINS-66771](https://issues.jenkins-ci.org/browse/JENKINS-66771) for detailed rationale.

The short take-away is: this PR adds an ability to run many agents (e.g. chroots or containers with various build/test environments) on a weaker host that can not execute all those environment at the same time - but can process them one at a time.

Practical use-case - many Linux containers with QEMU architectures, slow and resource consuming at run-time but can easily be pre-defined and stored on the same cloud host dedicated for a project.

Note that I sadly did not make any test-cases, just not proficient enough with Java to make some from scratch in finite time. Usually I would extend an existing test, but seems there is none for `inDemandDelay` and `idleDelay` either. So would have to defer this to anyone who can.

### Proposed changelog entries

* Add a "conflictsWith" concept to agents starting on demand

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] (If applicable) Jira issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [x] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

File: core/src/main/java/hudson/slaves/RetentionStrategy.java [hunk 1]
```diff
@@ -261,9 +307,14 @@ public long check(final SlaveComputer c) {
 
                 if (needComputer) {
                     // we've been in demand for long enough
-                    logger.log(Level.INFO, "Launching computer {0} as it has been in demand for {1}",
+                    if (hasConflict.size() > 0) {
```
Comment (timja) [inline core/src/main/java/hudson/slaves/RetentionStrategy.java:310]:
```suggestion
                    if (!hasConflict.isEmpty()) {
```

File: core/src/main/java/hudson/slaves/RetentionStrategy.java [hunk 2]
```diff
@@ -217,13 +238,38 @@ public long getIdleDelay() {
             return idleDelay;
         }
 
+        /**
+         * Getter for property 'conflictsWith'.
+         *
+         * @return Value for property 'conflictsWith'.
+         */
+        public String getConflictsWith() {
```
Comment (timja) [inline core/src/main/java/hudson/slaves/RetentionStrategy.java:246]: add a `doCheckConflictsWith` method to validate the regex?

File: core/src/main/java/hudson/slaves/RetentionStrategy.java [hunk 3]
```diff
@@ -193,10 +199,25 @@ public String getDisplayName() {
...
+         * declare each other as conflicting.
+         * This allows a Jenkins deployment to co-locate various pre-defined
+         * or otherwise provisioned build environments without overwhelming
+         * their host - and instead to permit round-robining, one by one.
+         */
+        private final String conflictsWith;
+
         @DataBoundConstructor
-        public Demand(long inDemandDelay, long idleDelay) {
+        public Demand(long inDemandDelay, long idleDelay, String conflictsWith) {
```
Comment (timja) [inline core/src/main/java/hudson/slaves/RetentionStrategy.java:217]: generally optional fields should be created by a setter (and annotated by `@DataboundSetter`

Comment (jimklimov) [reply core/src/main/java/hudson/slaves/RetentionStrategy.java]: fixed

File: core/src/main/java/hudson/slaves/RetentionStrategy.java (hunk 2 above)
Comment (jimklimov) [reply core/src/main/java/hudson/slaves/RetentionStrategy.java:250]: fixed

File: core/src/main/java/hudson/slaves/RetentionStrategy.java (hunk 1 above)
Comment (jimklimov) [reply core/src/main/java/hudson/slaves/RetentionStrategy.java]: fixed

**Comment by jimklimov** [comment]
Fixed the review comments (thanks @timja!) and some locally found imperfections.

Did not address auto-testing :\

**Review by daniel-beck — COMMENTED** [review]
Seems like this might be better off in a plugin implementing a more powerful `RetentionStrategy` than what is in core? The use case looks like it would be addressed well by this change, but seems a bit narrow for supporting it in core.

**Comment by jimklimov** [comment]
Gentle bump :)

This new strategy proved invaluable at the NUT CI farm, and needed no changes over the past month, so I'd love to not have to roll out forked builds but to use the vanilla upstream with this feature ;)

**Comment by daniel-beck** [comment]
Does https://github.com/daniel-beck/conflict-aware-ondemand-retention-strategy-plugin help?

**Comment by jimklimov** [comment]
Didn't try it yet, but logically looks cool! :)

If installed as plugin, would the tweak be configurable for the different
agent types (ssh, swarm, ...) as the current PR is?

Jim

**Comment by daniel-beck** [comment]
> If installed as plugin, would the tweak be configurable for the different agent types (ssh, swarm, ...) as the current PR is?

I expect so, agent launcher and retention strategy should be independent. As I wrote, I haven't tested in beyond making sure the UI works.

**Comment by jimklimov** [comment]
Sorry for prolonged delay with the end of year rush, but I finally got a moment to set that system up with your plugin version of my PR and a "vanilla" LTS Jenkins core (2.319.1).

Took some time to find the regexes to apply from backup of older configs and click up the configuration to use the new ondemand method in a couple dozen agents ;)

Finally queued some builds that should hit the QEMU VMs so will see by tomorrow how it goes.

**Comment by jimklimov** [comment]
@daniel-beck : as for repo "transfer", maybe better it could make way into jenkins-ci org and add me as maintainer to brush up the README, publish releases, etc?

I'll wait for the builds noted above to progress further (e.g. that it would hand over to next agents when initially started ones complete), but so far it looks promising that the feature can live outside the Jenkins core - thanks for the help!

**Comment by daniel-beck** [comment]
@jimklimov 

> as for repo "transfer", maybe better it could make way into jenkins-ci org and add me as maintainer to brush up the README, publish releases, etc?

Delete your fork and I'll transfer the repo to you, then you can request hosting. (Hosting works best when there are no forks to begin with.)

**Comment by daniel-beck** [comment]
Closing this PR for now given where this seems to be going.

**Comment by jimklimov** [comment]
> Delete your fork and I'll transfer the repo to you, then you can request hosting. (Hosting works best when there are no forks to begin with.)

Thanks, nuked :)

**Pull request closed without merging**